Fails to build on Windows with hdf5 1.10.5 from rtools40

Hi, we are trying to upgrade the Windows toolchain and libraries. This package fails to build with hdf5 1.10.5 on Windows. The cran package fails with an issue that looks like https://github.com/hhoeflin/hdf5r/issues/97:

```
*** arch - i386
cp 1_8_14/*.c 1_8_14/*.h .
/mingw32/bin/gcc  -I"C:/progra~1/R/R-testing/include" -DNDEBUG       -I/include/hdf5 -I/include/hdf5_hl -I/include/cmakeconf -I1_8_14 -I. -D__USE_MINGW_ANSI_STDIO   -O2 -Wall  -std=gnu99 -mtune=generic -c const_export.c -o const_export.o

/mingw32/bin/gcc  -I"C:/progra~1/R/R-testing/include" -DNDEBUG       -I/include/hdf5 -I/include/hdf5_hl -I/include/cmakeconf -I1_8_14 -I. -D__USE_MINGW_ANSI_STDIO   -O2 -Wall  -std=gnu99 -mtune=generic -c datatype_export.c -o datatype_ex
port.o
datatype_export.c: In function 'create_DT_H5D_mpio_no_collective_cause_t':
datatype_export.c:1471:12: error: 'H5D_MPIO_FILTERS' undeclared (first use in this function); did you mean 'H5Z_MAX_NFILTERS'?
   myenum = H5D_MPIO_FILTERS;
            ^~~~~~~~~~~~~~~~
            H5Z_MAX_NFILTERS
datatype_export.c:1471:12: note: each undeclared identifier is reported only once for each function it appears in
In file included from C:/rtools40/mingw32/lib/gcc/i686-w64-mingw32/8.3.0/include/stddef.h:1,
                 from C:/rtools40/mingw32/i686-w64-mingw32/include/inttypes.h:14,
                 from C:/rtools40/mingw32/include/H5public.h:54,
                 from C:/rtools40/mingw32/include/hdf5.h:22,
                 from global.h:23,
                 from datatype_export.h:23,
                 from datatype_export.c:20:
datatype_export.c: In function 'create_DT_H5F_info_t':
datatype_export.c:2272:41: error: 'H5F_info2_t' {aka 'struct H5F_info2_t'} has no member named 'super_ext_size'
   H5Tinsert(dtype_id, "super_ext_size", HOFFSET(H5F_info_t, super_ext_size), h5_datatype[DT_hsize_t]);
                                         ^~~~~~~
make: *** [C:/progra~1/R/R-testing/etc/i386/Makeconf:208: datatype_export.o] Error 1
ERROR: compilation failed for package 'hdf5r'

```

I also tried the Github release but that one fails even sooner:

```
*** arch - i386
/mingw32/bin/gcc  -I"C:/progra~1/R/R-testing/include" -DNDEBUG          -O2 -Wall  -std=gnu99 -mtune=generic -c H5Error.c -o H5Error.o
In file included from H5Error.c:19:
global.h:33:10: fatal error: HelperStructs.h: No such file or directory
 #include "HelperStructs.h"
          ^~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [C:/progra~1/R/R-testing/etc/i386/Makeconf:208: H5Error.o] Error 1
ERROR: compilation failed for package 'hdf5r'
* removing 'C:/R/library/hdf5r'
* restoring previous 'C:/R/library/hdf5r'
```

Can you please unblock us by looking into a fix?

have you changed the hdf5 version on windows?

On Wed, Oct 2, 2019 at 3:44 PM Jeroen Ooms <notifications@github.com> wrote:

> Hi, we are trying to upgrade the toolchain and this package fails to
> build. The cran package fails with an issue that looks like #97
> <https://github.com/hhoeflin/hdf5r/issues/97>:
>
> *** arch - i386
> cp 1_8_14/*.c 1_8_14/*.h .
> /mingw32/bin/gcc  -I"C:/progra~1/R/R-testing/include" -DNDEBUG       -I/include/hdf5 -I/include/hdf5_hl -I/include/cmakeconf -I1_8_14 -I. -D__USE_MINGW_ANSI_STDIO   -O2 -Wall  -std=gnu99 -mtune=generic -c const_export.c -o const_export.o
>
> /mingw32/bin/gcc  -I"C:/progra~1/R/R-testing/include" -DNDEBUG       -I/include/hdf5 -I/include/hdf5_hl -I/include/cmakeconf -I1_8_14 -I. -D__USE_MINGW_ANSI_STDIO   -O2 -Wall  -std=gnu99 -mtune=generic -c datatype_export.c -o datatype_ex
> port.o
> datatype_export.c: In function 'create_DT_H5D_mpio_no_collective_cause_t':
> datatype_export.c:1471:12: error: 'H5D_MPIO_FILTERS' undeclared (first use in this function); did you mean 'H5Z_MAX_NFILTERS'?
>    myenum = H5D_MPIO_FILTERS;
>             ^~~~~~~~~~~~~~~~
>             H5Z_MAX_NFILTERS
> datatype_export.c:1471:12: note: each undeclared identifier is reported only once for each function it appears in
> In file included from C:/rtools40/mingw32/lib/gcc/i686-w64-mingw32/8.3.0/include/stddef.h:1,
>                  from C:/rtools40/mingw32/i686-w64-mingw32/include/inttypes.h:14,
>                  from C:/rtools40/mingw32/include/H5public.h:54,
>                  from C:/rtools40/mingw32/include/hdf5.h:22,
>                  from global.h:23,
>                  from datatype_export.h:23,
>                  from datatype_export.c:20:
> datatype_export.c: In function 'create_DT_H5F_info_t':
> datatype_export.c:2272:41: error: 'H5F_info2_t' {aka 'struct H5F_info2_t'} has no member named 'super_ext_size'
>    H5Tinsert(dtype_id, "super_ext_size", HOFFSET(H5F_info_t, super_ext_size), h5_datatype[DT_hsize_t]);
>                                          ^~~~~~~
> make: *** [C:/progra~1/R/R-testing/etc/i386/Makeconf:208: datatype_export.o] Error 1
> ERROR: compilation failed for package 'hdf5r'
>
>
> I also tried the Github release but that one fails even sooner:
>
> *** arch - i386
> /mingw32/bin/gcc  -I"C:/progra~1/R/R-testing/include" -DNDEBUG          -O2 -Wall  -std=gnu99 -mtune=generic -c H5Error.c -o H5Error.o
> In file included from H5Error.c:19:
> global.h:33:10: fatal error: HelperStructs.h: No such file or directory
>  #include "HelperStructs.h"
>           ^~~~~~~~~~~~~~~~~
> compilation terminated.
> make: *** [C:/progra~1/R/R-testing/etc/i386/Makeconf:208: H5Error.o] Error 1
> ERROR: compilation failed for package 'hdf5r'
> * removing 'C:/R/library/hdf5r'
> * restoring previous 'C:/R/library/hdf5r'
>
> Can you please unblock us by looking into a fix?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/hhoeflin/hdf5r/issues/128?email_source=notifications&email_token=AASGMYVF3WEKZCKJQJ2RLC3QMSQTDA5CNFSM4I4WRDA2YY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4HPEIOBA>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AASGMYS22RAKVAHXB3FKVUTQMSQTDANCNFSM4I4WRDAQ>
> .
>

--- a/.Rbuildignore
+++ b/.Rbuildignore
@@ -28,3 +28,5 @@ wercker.yml
 ^docs$
 ^_pkgdown\.yml$
 ^cran-comments\.md$
+windows
+

--- a/Makefile
+++ b/Makefile
@@ -20,16 +20,6 @@ CHECKPATH := $(PKG_NAME).Rcheck
 CHECKLOG := `cat $(CHECKPATH)/00check.log`
 CURRENT_DIR := $(shell pwd)
 
-# Makevars used for CRAN release
-HDF5_VERSION_USE=1_8_14
-WINC=$${LIB_HDF5}/include
-WLIB=$${LIB_HDF5}/lib$${R_ARCH}
-HDF5R_CPPFLAGS=
-DEFINE_H5FREE_MEMORY=0
-HDF5R_LIBS=-L${WLIB} -lhdf5 -lhdf5_hl -lz -lm
-HDF5R_CFLAGS=-I${WINC}/hdf5 -I${WINC}/hdf5_hl -I${WINC}/cmakeconf -I${HDF5_VERSION_USE} -I.
-# Makevars end
-
 .PHONY: all build check manual install clean compileAttributes roxygen\
 	build-cran check-cran doc 
 
@@ -45,18 +35,6 @@ $(PKG_NAME)_$(PKG_VERSION).tar.gz: $(PKG_FILES)
 build-cran:
 	@make clean
 	@make roxygen
-	# Generate src/Makevars.win
-	sed -e 's#@HDF5_VERSION_USE@#${HDF5_VERSION_USE}#g' -e 's#@HDF5R_CFLAGS@#${HDF5R_CFLAGS}#g'  \
--e  's#@HDF5R_CPPFLAGS@#${HDF5R_CPPFLAGS}#g' -e  's#@HDF5R_LIBS@#${HDF5R_LIBS}#g' \
--e  's#@DEFINE_H5FREE_MEMORY@#${DEFINE_H5FREE_MEMORY}#g' \
-src/Makevars.in > src/Makevars.win
-	# Swap configure.win
-	@cp configure.win configure.win.temp; cat /dev/null >| configure.win
-	R CMD build --resave-data .
-	# Unswap configure.win
-	@cp configure.win.temp configure.win; rm configure.win.temp
-	# Remove src/Makevars.win
-	rm src/Makevars.win
 
 roxygen: $(R_FILES)
 	$(R) 'devtools::load_all(".", reset=TRUE, recompile = FALSE, export_all=FALSE)';

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -13,7 +13,6 @@ install:
 # Adapt as necessary starting from here
 
 build_script:
-  - Rscript -e "source('https://bioconductor.org/biocLite.R')"
   - travis-tool.sh install_deps
   - echo %CD%
   - ls
@@ -28,7 +27,6 @@ environment:
   WARNINGS_ARE_ERRORS: 1
   _R_CHECK_FORCE_SUGGESTS_: FALSE
   R_ARCH: x64
-  HDF5_VERSION_USE: 1_8_14
 
 artifacts:
   - path: '*.Rcheck\**\*.log'

--- a/src/Makevars.win
+++ b/src/Makevars.win
@@ -1,0 +1,26 @@
+WRAP = 1_10_3
+VERSION = 1.10.5
+LIB_HDF5 = ../windows/hdf5-$(VERSION)
+
+PKG_LIBS = -L${LIB_HDF5}/lib${R_ARCH} -lhdf5 -lhdf5_hl -lz -lm 
+PKG_CFLAGS=-I${LIB_HDF5}/include -I$(WRAP) -I. -D__USE_MINGW_ANSI_STDIO
+
+.PHONY: clean all winlibs
+
+all: clean winlibs $(SHLIB)
+
+OBJECTS = $(WRAP)/const_export.o $(WRAP)/datatype_export.o $(WRAP)/Wrapper_auto_H5A.o \
+	$(WRAP)/Wrapper_auto_H5.o $(WRAP)/Wrapper_auto_H5D.o $(WRAP)/Wrapper_auto_H5DS.o $(WRAP)/Wrapper_auto_H5E.o \
+	$(WRAP)/Wrapper_auto_H5F.o $(WRAP)/Wrapper_auto_H5G.o $(WRAP)/Wrapper_auto_H5I.o $(WRAP)/Wrapper_auto_H5IM.o \
+	$(WRAP)/Wrapper_auto_H5L.o $(WRAP)/Wrapper_auto_H5LT.o $(WRAP)/Wrapper_auto_H5O.o $(WRAP)/Wrapper_auto_H5P.o \
+	$(WRAP)/Wrapper_auto_H5R.o $(WRAP)/Wrapper_auto_H5S.o $(WRAP)/Wrapper_auto_H5TB.o $(WRAP)/Wrapper_auto_H5T.o \
+	$(WRAP)/Wrapper_auto_H5Z.o $(WRAP)/Wrapper_auto_H5FDcore.o $(WRAP)/Wrapper_auto_H5FDfamily.o \
+	$(WRAP)/Wrapper_auto_H5FDlog.o $(WRAP)/Wrapper_auto_H5FDsec2.o $(WRAP)/Wrapper_auto_H5FDstdio.o \
+	convert.o hdf5r_init.o H5Error.o H5ls.o Wrapper_manual_H5T.o
+
+winlibs:
+	@echo "Linking to HDF5 $(RWINLIB)"
+	"${R_HOME}/bin${R_ARCH_BIN}/Rscript.exe" "../tools/winlibs.R" $(VERSION)
+
+clean:
+	rm -f $(SHLIB) $(OBJECTS)

--- a/src/Makevars.win
+++ b/src/Makevars.win
@@ -1,8 +1,8 @@
-WRAP = 1_10_3
-VERSION = 1.10.5
+WRAP = 1_8_16
+VERSION = 1.8.16
 LIB_HDF5 = ../windows/hdf5-$(VERSION)
 
-PKG_LIBS = -L${LIB_HDF5}/lib${R_ARCH} -lhdf5 -lhdf5_hl -lz -lm 
+PKG_LIBS = -L${LIB_HDF5}/lib-4.9.3${R_ARCH} -lhdf5 -lhdf5_hl -lz -lm 
 PKG_CFLAGS=-I${LIB_HDF5}/include -I$(WRAP) -I. -D__USE_MINGW_ANSI_STDIO
 
 .PHONY: clean all winlibs

--- a/src/Makevars.win
+++ b/src/Makevars.win
@@ -1,13 +1,10 @@
 WRAP = 1_8_16
 VERSION = 1.8.16
-LIB_HDF5 = ../windows/hdf5-$(VERSION)
+RWINLIB = ../windows/hdf5-$(VERSION)
+PKG_LIBS = -L${RWINLIB}/lib-4.9.3${R_ARCH} -lhdf5 -lhdf5_hl -lz -lm 
+PKG_CPPFLAGS = -I${RWINLIB}/include -I$(WRAP) -I. -D__USE_MINGW_ANSI_STDIO
 
-PKG_LIBS = -L${LIB_HDF5}/lib-4.9.3${R_ARCH} -lhdf5 -lhdf5_hl -lz -lm 
-PKG_CFLAGS=-I${LIB_HDF5}/include -I$(WRAP) -I. -D__USE_MINGW_ANSI_STDIO
-
-.PHONY: clean all winlibs
-
-all: clean winlibs $(SHLIB)
+all: winlibs
 
 OBJECTS = $(WRAP)/const_export.o $(WRAP)/datatype_export.o $(WRAP)/Wrapper_auto_H5A.o \
 	$(WRAP)/Wrapper_auto_H5.o $(WRAP)/Wrapper_auto_H5D.o $(WRAP)/Wrapper_auto_H5DS.o $(WRAP)/Wrapper_auto_H5E.o \
@@ -22,5 +19,5 @@ winlibs:
 	@echo "Linking to HDF5 $(RWINLIB)"
 	"${R_HOME}/bin${R_ARCH_BIN}/Rscript.exe" "../tools/winlibs.R" $(VERSION)
 
-clean:
-	rm -f $(SHLIB) $(OBJECTS)
+.PHONY: all winlibs
+

--- a/tools/winlibs.R
+++ b/tools/winlibs.R
@@ -1,0 +1,9 @@
+# Build against rwinlib build of hdf5
+VERSION <- commandArgs(TRUE)
+if(!file.exists(sprintf("../windows/hdf5-%s/include/hdf5.h", VERSION))){
+  download.file(sprintf("https://github.com/rwinlib/hdf5/archive/v%s.zip", VERSION),
+                "lib.zip", quiet = TRUE)
+  dir.create("../windows", showWarnings = FALSE)
+  unzip("lib.zip", exdir = "../windows")
+  unlink("lib.zip")
+}

